### ISSUE TITLE
Switch to StatefulSet

### ISSUE THREAD

**Issue by nijel** [WeblateOrg/helm#54]
I saw the official Helm chart the other day and one thing stood out — it models Weblate as a stateless Deployment rather than StatefulSet which is great for stateful services. As far as I know, Weblate is currently a stateful service and can't be scaled horizontally. We started using Weblate on Kubernetes way before the official Helm chart got released and we first modelled it as Deployment too but the upgrades were somewhat problematic and it kept failing when trying to re-attach persistent disk to the newly spun up container. We used the "Recreate" rollout strategy but it would still fail and then we switched over to StatefulSet and this issue's been gone ever since. 

Anywya, the idea is, should we remodel Weblate as a StatefulService? Is there any specific reason why we're using the Deployment object? I'm assuming that you've already considered it and that there are some reasons that I might have not thought about.

_Originally posted by @mareksuscak in https://github.com/WeblateOrg/weblate/discussions/4806_

**Comment by Yann-J** [comment]
Good point... I can see the main pod has a volume attached, so indeed this wouldn't scale if we increase the replicas, unless the volume is mounted `ReadWriteMany`, so it seems the right way would indeed be to use a `StatefulSet` with each pod having their own volume.

However, when I look at what's persisted in this volume, it seems to be essentially the static result of the compilation step at startup... In that case, this volume doesn't really need to be persisted, it can very well be an `emptyDir` (just so it survives container crashes)...?

I think the only files that really need persistence are the `secret` file and `ssh` directory... but even then, they seem to be essentially read-only after an initial setup. This means they could be set up with a `pre-install` helm hook once, and then mounted as a single `ReadOnlyMany` volume by all pods of the same `Deployment`...

In general I think we should try to use Deployment as much as possible over StatefulSets, as every new volume has a cost...

**Comment by nijel** [comment]
It really contains data which is supposed to persist - user uploaded content (screenshots, fonts) or VCS data, see https://docs.weblate.org/en/latest/admin/config.html#data-dir for documentation. Some more insights on what is stored there is also available in https://github.com/WeblateOrg/weblate/issues/2984#issuecomment-526252515.

**Comment by mareksuscak** [comment]
Like @nijel pointed out above, the data directory does hold the user-generated data so `StatefulSet` would be more than appropriate. However, I'm not sure if Weblate can run multiple instances while maintaining consistency in user-generated data right now. In other words, would it correctly replicate all screenshots? Would each instance correctly synchronize all commits in a timely manner? I don't think we're quite there yet but please correct me if I'm wrong @nijel. That's the main reason for why this transition is on hold I'd say.

**Comment by nijel** [comment]
Yes, the filesystem has to be synchronous across Weblate instances

**Comment by Yann-J** [comment]
Ah yes of course... indeed, in this case, if replication is expected, unless the application manages it, switching to StatefulSet will not be enough.
I would suggest to simply mention that scaling (setting more than 1 replicas) is only possible with ReadWriteMany volumes. I see this accessMode and the storage class is already configurable in `values.yaml`. I'm not sure that using more than 1 replicas would be a very common use case anyway, as one instance should probably already be able to sustain a fair workload...
RWM volumes tend to be more expensive, so in this case we might want to limit it to the strict minimum of files that indeed have to be replicated across. Auto-generated statics probably shouldn't belong there (?).

**Comment by bartusz01** [comment]
Hi, we tested running multiple replicas, but ran into an issue, the css file is not always found (depending on which container the traffic is directed to), I suppose it can be fixed with session affinity, but it seems to be related to the these css files being located in `/app/cache/static/CACHE/css` which is not synced between the containers like `/app/data` directory.

**Comment by nijel** [comment]
You should run the same version in all replicas; otherwise things will break. /app/cache/static/CACHE/css is filled during container startup and does not need to be synced.

**Comment by bartusz01** [comment]
what version do you mean? Afaik all versions are equal between containers. 

FYI, the css file names are different between containers, when I run `ls /app/cache/static/CACHE/css`, I get e.g. an output like `output.82205c8x9f76.css  output.79f6539f66c2.css`, which is different in both containers, restarting a container will generate again different names. Meanwhile, in the browser I get a 404 on path "/static/CACHE/css/output.79f6539f66c2.css" if traffic is directed to the other container.

**Comment by bartusz01** [comment]
Thanks for the quick fix! 
Just to confirm, are you sure that it is safe to run multiple replicas (with RWX pv)? Nothing bad can happen with concurrent writes or file locks for instance?

**Comment by nijel** [comment]
Yes, it's safe. All file system accesses are lock protected using Redis, no file locks are used for that.

**Comment by zisuu** [comment]
Hi @nijel 

Thanks for the fix.

Do you maybe have an ETA until when [this commit](https://github.com/WeblateOrg/weblate/commit/90fbea8a41019ccd992f44f87c8268d42494823a) will be part of a new release? Currently we can not have Weblate HA on EKS because of this issue. With an active ChaosKube that randomly kills pods this is a nightmare. 😅

**Comment by nijel** [comment]
I've backported the patch to the Docker image in https://github.com/WeblateOrg/docker/commit/ec9086952ded6eb74394321086eefd89610f5a24, it will be available later today in `bleeding` and `edge` tags.

**Comment by zisuu** [comment]
is there any chance that this patch will also be released in a build with a version tag?

**Comment by nijel** [comment]
It was released in Weblate 4.18.2, so it's already there.

**Comment by zisuu** [comment]
Sorry I missed that. Awesome, thanks a lot

**Comment by zisuu** [comment]
fyi: switched to most recent helm chart and weblate version and it seems to work now. Can now run multiple replicas without running into this css bug anymore

**Comment by begemotik** [comment]
@bartusz01 @zisuu Hello guys!

Can you confirm that you still run multiple replicas using this helm chart with shared storage (EFS or other)? 

This documentation piece seem to be vague and have reference to docker-compose file where Weblate is being deployed in a distributed way (microservices). Your input would very helpful to asses self-hosted solution.
https://docs.weblate.org/en/latest/admin/install/docker.html#scaling-horizontally

Thank you.

**Comment by nijel** [comment]
The container still supports this, and I don't think the chart could break this.

**Comment by zisuu** [comment]
We run Weblate on AWS EKS with 2 replicas using the upstream helm chart (which still uses a Deployment, not a StatefulSet). 

Persistence is backed by EFS (AWS Elastic File System) with ReadWriteMany access mode, so both pods can mount the same volume simultaneously. This works fine for us.

Regarding horizontal scaling: Weblate does support it since the docker-compose based setup was added (the link you referenced). The key requirements are:

1. A shared filesystem that supports ReadWriteMany (e.g., EFS, NFS, CephFS) for the data volume
2. An external/managed database (not the bundled PostgreSQL container)
3. An external/managed Redis (or shared Redis)

The upstream chart still uses a Deployment rather than a StatefulSet, which works for us since EFS handles the shared storage. A StatefulSet would mainly help if you're using block storage (EBS/gp3) that only supports ReadWriteOnce -> in that case each replica would need its own volume, but then you'd lose the shared filesystem that Weblate needs for its git repositories.

**Comment by begemotik** [comment]
That's wonderful. Thank you for shedding light on the matter.

**Comment by begemotik** [comment]
@zisuu one catch though. 
I've reproduced the setup you mentioned and observe poor performance when adding new components (long initial repo scan) or making any repo maintenance actions. 

From the EFS perspective, we are not reaching any IOPS limits. I'm trying to figure out if there is anything I can do about it, or if EFS and Git repository are simply a bad match. Is this the case for you as well?

**Comment by begemotik** [comment]
UPD on EFS setup for shared storage:
```
weblate@weblate-6c6d5bfbcc-lcvvk:/app/data/vcs$ dd if=/dev/zero of=file2.bin bs=1M count=15000
15000+0 records in
15000+0 records out
15728640000 bytes (16 GB, 15 GiB) copied, 48.6663 s, 323 MB/s
```
I can write file of 15Gb size faster than Weblate clone a repo (58mb) with depth 1, when adding new component. It takes around 4 minutes:
<img width="375" height="126" alt="Image" src="https://github.com/user-attachments/assets/2ee740c3-6e4a-499f-bfd4-162411bf718d" />
